Convert TcpEcho.Client to F# and netstandard2.0

lukethenuke opened the pull request.

lukethenuke committed:
Convert TcpEcho.Client to F#

lukethenuke committed:
Replace netcoreapp2.1 target with netstandard2.0

knocte commented on src/Client/Client.fs:
```diff
@@ -0,0 +1,63 @@
+            let writer = writeToPipeAsync pipe.Writer socket |> Async.StartAsTask
+            let reader = readFromPipeAsync pipe.Reader "" |> Async.StartAsTask
+
+            do! [ writer :> Task; reader :> Task] |> Task.WhenAll |> Async.AwaitTask
```
do you mind using Async.Parallel instead of using Task.WhenAll please?

lukethenuke replied:
Another alternative would be to replace it with
```fsharp
let! _ = writeToPipeAsync pipe.Writer socket |> Async.StartChild
return! readFromPipeAsync pipe.Reader ""
```

knocte commented on src/Client/Client.fs:
```diff
@@ -0,0 +1,63 @@
+
+            do! [ writer :> Task; reader :> Task] |> Task.WhenAll |> Async.AwaitTask
+            return! Async.AwaitTask reader
+        } |> Async.StartAsTask
```
let's expose a method returning Async<T> and another one returning Task<T> (which calls to the former by using Async.StartAsTask). This way we would have F#-friendly API and C#-friendly API.

lukethenuke replied:
Added CallAsyncAsTask, if you have a better naming suggestion tell me :)

knocte commented on src/Client/Client.fs:
```diff
@@ -0,0 +1,63 @@
+
+    let rec readFromPipeAsync (reader: PipeReader) str = async {
+        let! result = reader.ReadAsync().AsTask() |> Async.AwaitTask
+        let buffer : ReadOnlySequence<byte> = result.Buffer
```
nit(cosmetic): you used no spaces before colons in other places, so let's be consistent in this line above too

Commit:
Use Async.Parallel instead of Task and add C#-friendly method

knocte committed:
sln: update

knocte commented:
Are you on Windows or macOS? The tests project doesn't compile for me in VS4Mac :'(

lukethenuke committed:
Fix StratumClient implementation

lukethenuke commented:
Yeah, the upstream clients needs to be updated. Does this work better?

lukethenuke committed:
Fix projectreference to StratumClient

lukethenuke commented:
The tests project should probably reference the StratumClient too :)

knocte commented:
Mmm, weird, I still get a compiler error! That's why I was asking what IDE are you using?

On macOS (VS4Mac 8.1) I get:

```
/Users/knocte/Documents/Code/tcpEchoLuke/src/StratumClient/FSC: Error FS0078: Unable to find the file '/Users/knocte/Documents/Code/tcpEchoLuke/src/Client/bin/Debug/netstandard2.0/Client.dll' in any of /Users/knocte/.nuget/packages/netstandard.library/2.0.3/build/netstandard2.0/ref /Users/knocte/Documents/Code/tcpEchoLuke/src/StratumClient /Users/knocte/.nuget/packages/fsharp.core/4.2.3/lib/netstandard1.6 (FS0078) (StratumClient)
```

lukethenuke commented:
I use the dotnet cli (2.1.403) to build and run on osx. 
The net461 is built with dotnet build on osx and run on a windows machine.
The tests are checked with dotnet tests.
I can also run and test it on windows using the dotnet cli.

It seems like it can not find Client.dll, firstly look in the the bin/Debug/netstandard2.0 and see if you can find the binary, if not try building only the Client library first.
Could it be compatibilty issues with netstandard and vs4mac? What if you add net461; to StratumClient?

knocte committed:
Fix build

knocte committed:
Rework exception handling to expose networking bug

One example of the bug:
System.Exception : Problem deserializing <<<<<<<<0000232200203a4e283d1e0b48e056678642f6e5bfe8151db0658f4c6fbad859fe02a9d7b69fffffffffe0136c03a9f573efc88371d818930be81a016e9648bc047dac4f9ac7a93f83a200000000232200203edb040081ee8fceed689e31c79cb67e781d33e85758f903edcd70ef4b75001dffffffff55c7cbe3c6a4233fd9aaabcd53b9013a0a61f6be8d70e3e5e54fec631770aeca00000000232200207b8b7ef82e696d835b06574476d46a683aa32f90cd8c4c6f5471282d7e698c0dffffffffd106ced87c0d350c9a4cb8b1ba69bf16428092b4c06898b6f9280f9a7396ce0904010000232200200fa4eeff73da7cbb7326ad57b7b9040b220e49f30b00f491f04b2fa0bc1d2804ffffffff0cf8162d000000000017a9145131075257d8b8de8298e7c52891eb4b87823b93873000c9010000000017a9140c1b069ba6d0691325cf15acf6b9e7c67f57b4c487b0feea0b0000000017a914e9210a740da86bc09c57045d9c53d03457a0209d8730208e030000000017a914e682c1068784415413f3feefa467e39da8fef3a187300b341d0000000017a914021f420dd06de86c7e82400e391e9c0d26423d518784233556000000001976a91466e1877107b93295873d7c607a0104c60064d38f88ac60df4000000000001976a91470f73e2ce9ebeadfbaf871e9b41b8ec3e56e9a7f88ac700ebd000000000016001449eb0000232200203a4e283d1e0b48e056678642f6e5bfe8151db0658f4c6fbad859fe02a9d7b69fffffffffe0136c03a9f573efc88371d818930be81a016e9648bc047dac4f9ac7a93f83a200000000232200203edb040081ee8fceed689e31c79cb67e781d33e85758f903edcd70ef4b75001dffffffff55c7cbe3c6a4233fd9aaabcd53b9013a0a61f6be8d70e3e5e54fec631770aeca00000000232200207b8b7ef82e696d835b06574476d46a683aa32f90cd8c4c6f5471282d7e698c0dffffffffd106ced87c0d350c9a4cb8b1ba69bf16428092b4c06898b6f9280f9a7396ce0904010000232200200fa4eeff73da7cbb7326ad57b7b9040b220e49f30b00f491f04b2fa0bc1d2804ffffffff0cf8162d000000000017a9145131075257d8b8de8298e7c52891eb4b87823b93873000c9010000000017a9140c1b069ba6d0691325cf15acf6b9e7c67f57b4c487b0feea0b0000000017a914e9210a740da86bc09c57045d9c53d03457a0209d8730208e030000000017a914e682c1068784415413f3feefa467e39da8fef3a187300b341d0000000017a914021f420dd06de86c7e82400e391e9c0d26423d518784233556000000001976a91466e1877107b93295873d7c607a0104c60064d38f88ac60df4000000000001976a91470f73e2ce9ebeadfbaf871e9b41b8ec3e56e9a7f88ac700ebd000000000016001449eb23af00ad1fa0cf32c6b52d92521b8c48f90be8b48c030000000017a9143003f822030a84b19d6cef5edab891a1ac6510128760b571000000000017a9147db9abfc8fce4da737a7108a030506d20c11acff87704a36000000000017a9147c7cedb301723552d1181460ea32f150dace05e9871adc762d0000000017a9144ad47d293d97e4d5cdfcbbdd4c3aa1453a9d48a487040047304402201facc7a954a6451b76a7e8c63b155280f5de9cbd48da6277cb40b26ace4e96de02203ccdf68c3bd917c6d5e85c642fc96824f380db7dbc726f02c52291de07226a350148304502210081622e0cda8b179dab0eeea68523f3cb1acdc98aaa84b860a7c1630a8073524d0220448044dca6a59267bb25bab67e4f827c91edc1f6c9675b2e1d293c7f4bf3e5c70169522102f72066131d60001b7d7abf39c6300aeb9f8cef16cb4fb9f238ae815111a73c49210285b875ddea788d7dae5b9d4d0a3992c65396c4c9d2fc710d4de74a21d8febbea210269a6a07f1168b5e2ae32844a8df637b23acf6a41f930e67b03b2b712dff6603453ae04004730440220750789f6ff389423fc6c2929341cdc55208798550ed722b718879617f15bf267022068f30b2c7d17c064e7e18856208b364bf53f0e9e1684a1165dd6f5ca94d98a5d01473044022059e9639cd64fe0c25c49a81993871aea9b5be81cf50d3507a685d84b5d2e153702205f4fb28081096f24f00686d23819d75f3c7f288dec7f9622d38b8d2172a63f710169522102952139ccb86b16383b29039bf76febbbfb1c41c15ec631c1b13f1424f7680a9321021c44aa55510b353e397f9f2b3e7dee36282f1421651fef96c31ba612c11627462103bbf6336d9f021ae4ee642af75156b644195dbe23a8de5cb62600b58e139d242653ae0400473044022070c844b19f917c16fc644957bafb3308b2cb244cdbce48ab763302d290272ef002204f6ca015707a9ffcbc3bee63c66a035430a7988495ce2f9c1d8e22f81c9ab0890147304402200ece13be1073c262d5a86c8078ce16fe6f2fa456665e6284eddd42b42006f74e022074fa8226a4ca112483736a9f0b47800b99197ffc8df3bcce6c3bef268a2a526a0169522102216a2779be169379d783d25180040e079da9193770e63bd5de8dc06c5834c0bd2102da354fb8873567ae83fef73a4fa783448a9b481e7f69619a20983149b1f19b402102e64c3c1e222b9cbdfb7f70b5bb011b89e0642eb7cfc94f2750948aeb54f8046553ae04004730440220431cde3ecb90c38c7f1bbafb4a417a8183e314eecc8a374fc67d5847b6822fd102201ff9e2ecc57bcdfb69ba1a9f78d132d7123b37e6838c7e77e37278b6269384e301483045022100df2ba7d95b973d30871a90f35757ad13b61c9f2deb8e36b675d19de4d5dfd12f02203f3b5dadd71304080a63c4031f4a6d7b30f3f742fc44e392a69cca947159d0890169522103630ae85a979ee329fa729b361a114254da4e871bce8629e39c26a26fb6ab85e82102d566976b3c356c2062a75e938ee538edef5ced6f6f8ca827c3b004b6c5f256a42102f5de580abf9fd69275cfc105ba165664d1d79b06f3aae1f512d778624c9ef8cb53ae0400473044022064b121d06f145acd2e84cf6ae73c6c8ba8119ac9450a93c9a8fcb882c98a7bd102206a5e8ff6b4b16b212d3c1cc92864ef9b2c804bc26f268399799a36414664923c0147304402200a99894da1fad45c7d62aceb0aa9d7a05f4304e4498327d8827f308ed8e70f43022034f68f701c3bb2afedc69ab7ba402a0d2277ef530fd89376836bb55151cf50540169522103ab69777a57f9cce9f3eea8079a2a1c89c07f9068022bf80686242efa1b60bbe021023718e1ca5f2f6e38841b627a8fa5c9a66f767d6a05367508cea960f21bb2feea2102e56e70e1cfac75b661131ac924aee909920d948ffb08f879bfa5ea1d3fecb9d353ae04004830450221008244512d55c6841c77a4d126fa5a33c92c247bf61d47c7dca0a7256ae8212e5f022019838e0aa334efa4a590b7398f3fdf6c2ef2a3d00f3856814ed04a9a0e56e54c01483045022100a8972a7a9a6852b0f96de6fa0fba67748551591ff9bec42a0df7d1f0a03392670220254d76e53a9c777aedbd2937c2f45324ff88a3ea15103bf9e325bd336d6cb2f501695221027d0f541fdf168131c9003d824c486062a9b6869f000cd65639a76255d05c3ee72103e86ad41675d5fcb7ce4d4b60e8f2ad122f873239d514bf1bceca49b74536a8aa2102805e8f635752cf9a42c38bb80f224b3212882e6c3c50bebba52eb7ab0935cae653ae040047304402202c71af53522c21e981863a336fd8b98394a981dbf8ac4a75c7b4f48c924faf56022057ef76dfcb327d39b4bc79ca5d3d604522ecfae6ee434f943ea714cf5ba0593301473044022068282d81f9968ef31db541e9b72aefff62aefd889e12c97f61e41e20d57d715d02202f54462beb8c4a462583c4917c3d937c036c5c5983c84c5be132cb9bef8d908c0169522102f116a19e2900371ae7f4be5523d9845654dedbce090e2fa30e816732d5d8251d2102d6f4005dfc28cb3c0def8c1f52f2d23d486dfbfa3ffa43c94a4b3d43c920c93e2103c6fac560735d78d95cd63652675e643c8e0433002679cd032143d8411dba6b1153ae0400483045022100a6bdc5206380571da68cde53f1ef19dca58c4d2d9cbc4220fe7e12e70fc9b0bd022011cad80c2bf83f85c111d5d3891f634a393c7ca57513ff2b5813e87bc85ae50501473044022100f598f696af768e2290f46873c3873876659fe60fe73c0068136e746a8f3327ef021f1b9de6928c8cf1245605ae705d7d1304efbbd385d4b77a81ab66bf8dda6a9401695221038acc7e5bd34e1198341f6c5561eba4580ce930ebfd78e9705fa51aaa34b8035b2103a40b87c3a637f4ab9b4cfadcf0537bfb71753b4f070e896eeeb0615bb21024f12103dbbc3ab658aef4b292042c3e99de6affcf573249a40bd6c225e4d84168bd01b253ae0400473044022041661469f59e7a72a9e1835323397fd4920e3a234c5c346afe4974c589ae25ee02207fe4b1c42b13dec0e1372446db604f8bda2be95c9bf0aa2c26dad8530ad11170014730440220417ce3b947dc37368af439c0afd347bcbbfed55b769e29bed1947079671197bf022079747806204ce6828a8d6b34703781f049124c39f8385fb1d1a880d9325e98850169522102aee5147d14dc3aa1da07594703864035112c1ce140ea7955599e9c3b911d65bc21022942c8ffc8271de9e763a3299bfc919b4077455c010aec920abd085df6ab73a621022c3f85c185fd927ed3e81f02ccaae1d45720c0929a0d1a6e9accd0c8d04342ef53ae040047304402206710fd815cbc1aaefae6fbc4b603f4f8b96516ab9f5dd69f371692ff0c2bfd34022040c6aebb39386aff976bb468b0c71e350a378b5a3847ccc9947ee522bb687c4e0147304402202cf317b968ed838f19634f5ea56ec0a521bbefeffe7a4c9498899d2b23fff4d5022009610c7bbd95f7cb212ee07299edfbb56c6ac8bf796b14ae9f2f64d5a08b79e30169522103ab69777a57f9cce9f3eea8079a2a1c89c07f9068022bf80686242efa1b60bbe021023718e1ca5f2f6e38841b627a8fa5c9a66f767d6a05367508cea960f21bb2feea2102e56e70e1cfac75b661131ac924aee909920d948ffb08f879bfa5ea1d3fecb9d353ae00000000", "id": 17}>>>>>>>>

knocte commented:
> What if you add net461; to StratumClient?

That fixed it, thanks! (I pushed this little fix)

But I also pushed a new commit to rework the exception handling that will make you see that there is a bug in the networking code. It's a flaky bug because it cannot be reproduced 100% of the times but if you run the `ConnectWithElectrumServersTransactionGet` test many times, you will see it.

An example stacktrace of the bug:
```
System.Exception : Problem deserializing <<<<<<<<0000232200203a4e283d1e0b48e056678642f6e5bfe8151db0658f4c6fbad859fe02a9d7b69fffffffffe0136c03a9f573efc88371d818930be81a016e9648bc047dac4f9ac7a93f83a200000000232200203edb040081ee8fceed689e31c79cb67e781d33e85758f903edcd70ef4b75001dffffffff55c7cbe3c6a4233fd9aaabcd53b9013a0a61f6be8d70e3e5e54fec631770aeca00000000232200207b8b7ef82e696d835b06574476d46a683aa32f90cd8c4c6f5471282d7e698c0dffffffffd106ced87c0d350c9a4cb8b1ba69bf16428092b4c06898b6f9280f9a7396ce0904010000232200200fa4eeff73da7cbb7326ad57b7b9040b220e49f30b00f491f04b2fa0bc1d2804ffffffff0cf8162d000000000017a9145131075257d8b8de8298e7c52891eb4b87823b93873000c9010000000017a9140c1b069ba6d0691325cf15acf6b9e7c67f57b4c487b0feea0b0000000017a914e9210a740da86bc09c57045d9c53d03457a0209d8730208e030000000017a914e682c1068784415413f3feefa467e39da8fef3a187300b341d0000000017a914021f420dd06de86c7e82400e391e9c0d26423d518784233556000000001976a91466e1877107b93295873d7c607a0104c60064d38f88ac60df4000000000001976a91470f73e2ce9ebeadfbaf871e9b41b8ec3e56e9a7f88ac700ebd000000000016001449eb0000232200203a4e283d1e0b48e056678642f6e5bfe8151db0658f4c6fbad859fe02a9d7b69fffffffffe0136c03a9f573efc88371d818930be81a016e9648bc047dac4f9ac7a93f83a200000000232200203edb040081ee8fceed689e31c79cb67e781d33e85758f903edcd70ef4b75001dffffffff55c7cbe3c6a4233fd9aaabcd53b9013a0a61f6be8d70e3e5e54fec631770aeca00000000232200207b8b7ef82e696d835b06574476d46a683aa32f90cd8c4c6f5471282d7e698c0dffffffffd106ced87c0d350c9a4cb8b1ba69bf16428092b4c06898b6f9280f9a7396ce0904010000232200200fa4eeff73da7cbb7326ad57b7b9040b220e49f30b00f491f04b2fa0bc1d2804ffffffff0cf8162d000000000017a9145131075257d8b8de8298e7c52891eb4b87823b93873000c9010000000017a9140c1b069ba6d0691325cf15acf6b9e7c67f57b4c487b0feea0b0000000017a914e9210a740da86bc09c57045d9c53d03457a0209d8730208e030000000017a914e682c1068784415413f3feefa467e39da8fef3a187300b341d0000000017a914021f420dd06de86c7e82400e391e9c0d26423d518784233556000000001976a91466e1877107b93295873d7c607a0104c60064d38f88ac60df4000000000001976a91470f73e2ce9ebeadfbaf871e9b41b8ec3e56e9a7f88ac700ebd000000000016001449eb23af00ad1fa0cf32c6b52d92521b8c48f90be8b48c030000000017a9143003f822030a84b19d6cef5edab891a1ac6510128760b571000000000017a9147db9abfc8fce4da737a7108a030506d20c11acff87704a36000000000017a9147c7cedb301723552d1181460ea32f150dace05e9871adc762d0000000017a9144ad47d293d97e4d5cdfcbbdd4c3aa1453a9d48a487040047304402201facc7a954a6451b76a7e8c63b155280f5de9cbd48da6277cb40b26ace4e96de02203ccdf68c3bd917c6d5e85c642fc96824f380db7dbc726f02c52291de07226a350148304502210081622e0cda8b179dab0eeea68523f3cb1acdc98aaa84b860a7c1630a8073524d0220448044dca6a59267bb25bab67e4f827c91edc1f6c9675b2e1d293c7f4bf3e5c70169522102f72066131d60001b7d7abf39c6300aeb9f8cef16cb4fb9f238ae815111a73c49210285b875ddea788d7dae5b9d4d0a3992c65396c4c9d2fc710d4de74a21d8febbea210269a6a07f1168b5e2ae32844a8df637b23acf6a41f930e67b03b2b712dff6603453ae04004730440220750789f6ff389423fc6c2929341cdc55208798550ed722b718879617f15bf267022068f30b2c7d17c064e7e18856208b364bf53f0e9e1684a1165dd6f5ca94d98a5d01473044022059e9639cd64fe0c25c49a81993871aea9b5be81cf50d3507a685d84b5d2e153702205f4fb28081096f24f00686d23819d75f3c7f288dec7f9622d38b8d2172a63f710169522102952139ccb86b16383b29039bf76febbbfb1c41c15ec631c1b13f1424f7680a9321021c44aa55510b353e397f9f2b3e7dee36282f1421651fef96c31ba612c11627462103bbf6336d9f021ae4ee642af75156b644195dbe23a8de5cb62600b58e139d242653ae0400473044022070c844b19f917c16fc644957bafb3308b2cb244cdbce48ab763302d290272ef002204f6ca015707a9ffcbc3bee63c66a035430a7988495ce2f9c1d8e22f81c9ab0890147304402200ece13be1073c262d5a86c8078ce16fe6f2fa456665e6284eddd42b42006f74e022074fa8226a4ca112483736a9f0b47800b99197ffc8df3bcce6c3bef268a2a526a0169522102216a2779be169379d783d25180040e079da9193770e63bd5de8dc06c5834c0bd2102da354fb8873567ae83fef73a4fa783448a9b481e7f69619a20983149b1f19b402102e64c3c1e222b9cbdfb7f70b5bb011b89e0642eb7cfc94f2750948aeb54f8046553ae04004730440220431cde3ecb90c38c7f1bbafb4a417a8183e314eecc8a374fc67d5847b6822fd102201ff9e2ecc57bcdfb69ba1a9f78d132d7123b37e6838c7e77e37278b6269384e301483045022100df2ba7d95b973d30871a90f35757ad13b61c9f2deb8e36b675d19de4d5dfd12f02203f3b5dadd71304080a63c4031f4a6d7b30f3f742fc44e392a69cca947159d0890169522103630ae85a979ee329fa729b361a114254da4e871bce8629e39c26a26fb6ab85e82102d566976b3c356c2062a75e938ee538edef5ced6f6f8ca827c3b004b6c5f256a42102f5de580abf9fd69275cfc105ba165664d1d79b06f3aae1f512d778624c9ef8cb53ae0400473044022064b121d06f145acd2e84cf6ae73c6c8ba8119ac9450a93c9a8fcb882c98a7bd102206a5e8ff6b4b16b212d3c1cc92864ef9b2c804bc26f268399799a36414664923c0147304402200a99894da1fad45c7d62aceb0aa9d7a05f4304e4498327d8827f308ed8e70f43022034f68f701c3bb2afedc69ab7ba402a0d2277ef530fd89376836bb55151cf50540169522103ab69777a57f9cce9f3eea8079a2a1c89c07f9068022bf80686242efa1b60bbe021023718e1ca5f2f6e38841b627a8fa5c9a66f767d6a05367508cea960f21bb2feea2102e56e70e1cfac75b661131ac924aee909920d948ffb08f879bfa5ea1d3fecb9d353ae04004830450221008244512d55c6841c77a4d126fa5a33c92c247bf61d47c7dca0a7256ae8212e5f022019838e0aa334efa4a590b7398f3fdf6c2ef2a3d00f3856814ed04a9a0e56e54c01483045022100a8972a7a9a6852b0f96de6fa0fba67748551591ff9bec42a0df7d1f0a03392670220254d76e53a9c777aedbd2937c2f45324ff88a3ea15103bf9e325bd336d6cb2f501695221027d0f541fdf168131c9003d824c486062a9b6869f000cd65639a76255d05c3ee72103e86ad41675d5fcb7ce4d4b60e8f2ad122f873239d514bf1bceca49b74536a8aa2102805e8f635752cf9a42c38bb80f224b3212882e6c3c50bebba52eb7ab0935cae653ae040047304402202c71af53522c21e981863a336fd8b98394a981dbf8ac4a75c7b4f48c924faf56022057ef76dfcb327d39b4bc79ca5d3d604522ecfae6ee434f943ea714cf5ba0593301473044022068282d81f9968ef31db541e9b72aefff62aefd889e12c97f61e41e20d57d715d02202f54462beb8c4a462583c4917c3d937c036c5c5983c84c5be132cb9bef8d908c0169522102f116a19e2900371ae7f4be5523d9845654dedbce090e2fa30e816732d5d8251d2102d6f4005dfc28cb3c0def8c1f52f2d23d486dfbfa3ffa43c94a4b3d43c920c93e2103c6fac560735d78d95cd63652675e643c8e0433002679cd032143d8411dba6b1153ae0400483045022100a6bdc5206380571da68cde53f1ef19dca58c4d2d9cbc4220fe7e12e70fc9b0bd022011cad80c2bf83f85c111d5d3891f634a393c7ca57513ff2b5813e87bc85ae50501473044022100f598f696af768e2290f46873c3873876659fe60fe73c0068136e746a8f3327ef021f1b9de6928c8cf1245605ae705d7d1304efbbd385d4b77a81ab66bf8dda6a9401695221038acc7e5bd34e1198341f6c5561eba4580ce930ebfd78e9705fa51aaa34b8035b2103a40b87c3a637f4ab9b4cfadcf0537bfb71753b4f070e896eeeb0615bb21024f12103dbbc3ab658aef4b292042c3e99de6affcf573249a40bd6c225e4d84168bd01b253ae0400473044022041661469f59e7a72a9e1835323397fd4920e3a234c5c346afe4974c589ae25ee02207fe4b1c42b13dec0e1372446db604f8bda2be95c9bf0aa2c26dad8530ad11170014730440220417ce3b947dc37368af439c0afd347bcbbfed55b769e29bed1947079671197bf022079747806204ce6828a8d6b34703781f049124c39f8385fb1d1a880d9325e98850169522102aee5147d14dc3aa1da07594703864035112c1ce140ea7955599e9c3b911d65bc21022942c8ffc8271de9e763a3299bfc919b4077455c010aec920abd085df6ab73a621022c3f85c185fd927ed3e81f02ccaae1d45720c0929a0d1a6e9accd0c8d04342ef53ae040047304402206710fd815cbc1aaefae6fbc4b603f4f8b96516ab9f5dd69f371692ff0c2bfd34022040c6aebb39386aff976bb468b0c71e350a378b5a3847ccc9947ee522bb687c4e0147304402202cf317b968ed838f19634f5ea56ec0a521bbefeffe7a4c9498899d2b23fff4d5022009610c7bbd95f7cb212ee07299edfbb56c6ac8bf796b14ae9f2f64d5a08b79e30169522103ab69777a57f9cce9f3eea8079a2a1c89c07f9068022bf80686242efa1b60bbe021023718e1ca5f2f6e38841b627a8fa5c9a66f767d6a05367508cea960f21bb2feea2102e56e70e1cfac75b661131ac924aee909920d948ffb08f879bfa5ea1d3fecb9d353ae00000000", "id": 17}>>>>>>>>
```

As you can see, it's not gathering properly all the content from the server.

Commit:
Fix partial response on slow sockets

lukethenuke commented:
The bug is the same as in the C# code, if there is a pause on the network after the server flush, the available bytes reported will be 0 and the reader will be closed. 
This commits fixes that, so it will wait for the socket to properly close. (Which means we will need to add the timeout back aswell)

knocte commented:
> The bug is the same as in the C# code

Before I gitpulled your last commit, I actually realised that:
a) the C# code had a flaky bug, true, but
b) your code was always failing (so, deterministic bug, 100% reproducible, not flaky)

I've gitpulled your code now and, while it looks correct, somehow doesn't make the test pass, it seems to hang. I've increased the timeout just in case many servers were being slow, to no avail. Did you run the test yourself?

lukethenuke commented:
No i tried it against a local socket server to be able to inser delays between flushes, but i tried it now and it hangs for me aswell (on the non-tor services). Will look into that.

Commit:
Fix timeout that hangs when the server does end the socket

lukethenuke commented:
Ok, so it turns out some servers does not end the socket after sending it's response. This fix timeout the read correctly now, it also throws the data as an IncompleteResponseException upwards to the StratumClient since the data may be fine so it gets an last chance to parse

knocte commented:
I think we're almost there! Only 3 things to address:

1. You added the comment "AggregateException so can't catch TimeoutException". Well, you could check that the innerexception is of type TimeoutException, otherwise raise. I prefer this because I hate typeless catches. Do you mind changing it?

2. When running the test I caught a NRE, which I think is because in some cases the IncompleteResponse is really not wrapping anything of value. Maybe you should just wrap the TimeoutException in this case with a ConnectionUnsuccessfulException to make the tests ignore this server. FTR this was the exception:

```
System.NullReferenceException : Object reference not set to an instance of an object
Stack trace:
  at Client.Tests.ClientTest.ConnectWithElectrumServersTransactionGet () [0x000cc] in /Users/knocte/Documents/Code/tcpEchoLuke/src/Client.Tests/ClientTests.cs:46 
  at NUnit.Framework.Internal.ExceptionHelper.Rethrow (System.Exception exception) [0x00006] in <64cd73ddfc8a470aa513c8243b429bdb>:0 
  at NUnit.Framework.Internal.AsyncInvocationRegion+AsyncTaskInvocationRegion.WaitForPendingOperationsToComplete (System.Object invocationResult) [0x00030] in <64cd73ddfc8a470aa513c8243b429bdb>:0 
  at NUnit.Framework.Internal.Commands.TestMethodCommand.RunAsyncTestMethod (NUnit.Framework.Internal.TestExecutionContext context) [0x00038] in <64cd73ddfc8a470aa513c8243b429bdb>:0 
```

3. If you run the tests several times you will notice that there's still a flaky bug, which I think is not related to timeouts or slow servers, because the json collected by the client code is incomplete (but incomplete from the left side, not incomplete from the right side, I mean, you can see the end of the string is the end of a json string, but the beginning of it is truncated, see below):

```
System.Exception : Problem deserializing <<<<<<<<0000232200203a4e283d1e0b48e056678642f6e5bfe8151db0658f4c6fbad859fe02a9d7b69fffffffffe0136c03a9f573efc88371d818930be81a016e9648bc047dac4f9ac7a93f83a200000000232200203edb040081ee8fceed689e31c79cb67e781d33e85758f903edcd70ef4b75001dffffffff55c7cbe3c6a4233fd9aaabcd53b9013a0a61f6be8d70e3e5e54fec631770aeca00000000232200207b8b7ef82e696d835b06574476d46a683aa32f90cd8c4c6f5471282d7e698c0dffffffffd106ced87c0d350c9a4cb8b1ba69bf16428092b4c06898b6f9280f9a7396ce0904010000232200200fa4eeff73da7cbb7326ad57b7b9040b220e49f30b00f491f04b2fa0bc1d2804ffffffff0cf8162d000000000017a9145131075257d8b8de8298e7c52891eb4b87823b93873000c9010000000017a9140c1b069ba6d0691325cf15acf6b9e7c67f57b4c487b0feea0b0000000017a914e9210a740da86bc09c57045d9c53d03457a0209d8730208e030000000017a914e682c1068784415413f3feefa467e39da8fef3a187300b341d0000000017a914021f420dd06de86c7e82400e391e9c0d26423d518784233556000000001976a91466e1877107b93295873d7c607a0104c60064d38f88ac60df4000000000001976a91470f73e2ce9ebeadfbaf871e9b41b8ec3e56e9a7f88ac700ebd000000000016001449eb0000232200203a4e283d1e0b48e056678642f6e5bfe8151db0658f4c6fbad859fe02a9d7b69fffffffffe0136c03a9f573efc88371d818930be81a016e9648bc047dac4f9ac7a93f83a200000000232200203edb040081ee8fceed689e31c79cb67e781d33e85758f903edcd70ef4b75001dffffffff55c7cbe3c6a4233fd9aaabcd53b9013a0a61f6be8d70e3e5e54fec631770aeca00000000232200207b8b7ef82e696d835b06574476d46a683aa32f90cd8c4c6f5471282d7e698c0dffffffffd106ced87c0d350c9a4cb8b1ba69bf16428092b4c06898b6f9280f9a7396ce0904010000232200200fa4eeff73da7cbb7326ad57b7b9040b220e49f30b00f491f04b2fa0bc1d2804ffffffff0cf8162d000000000017a9145131075257d8b8de8298e7c52891eb4b87823b93873000c9010000000017a9140c1b069ba6d0691325cf15acf6b9e7c67f57b4c487b0feea0b0000000017a914e9210a740da86bc09c57045d9c53d03457a0209d8730208e030000000017a914e682c1068784415413f3feefa467e39da8fef3a187300b341d0000000017a914021f420dd06de86c7e82400e391e9c0d26423d518784233556000000001976a91466e1877107b93295873d7c607a0104c60064d38f88ac60df4000000000001976a91470f73e2ce9ebeadfbaf871e9b41b8ec3e56e9a7f88ac700ebd000000000016001449eb23af00ad1fa0cf32c6b52d92521b8c48f90be8b48c030000000017a9143003f822030a84b19d6cef5edab891a1ac6510128760b571000000000017a9147db9abfc8fce4da737a7108a030506d20c11acff87704a36000000000017a9147c7cedb301723552d1181460ea32f150dace05e9871adc762d0000000017a9144ad47d293d97e4d5cdfcbbdd4c3aa1453a9d48a487040047304402201facc7a954a6451b76a7e8c63b155280f5de9cbd48da6277cb40b26ace4e96de02203ccdf68c3bd917c6d5e85c642fc96824f380db7dbc726f02c52291de07226a350148304502210081622e0cda8b179dab0eeea68523f3cb1acdc98aaa84b860a7c1630a8073524d0220448044dca6a59267bb25bab67e4f827c91edc1f6c9675b2e1d293c7f4bf3e5c70169522102f72066131d60001b7d7abf39c6300aeb9f8cef16cb4fb9f238ae815111a73c49210285b875ddea788d7dae5b9d4d0a3992c65396c4c9d2fc710d4de74a21d8febbea210269a6a07f1168b5e2ae32844a8df637b23acf6a41f930e67b03b2b712dff6603453ae04004730440220750789f6ff389423fc6c2929341cdc55208798550ed722b718879617f15bf267022068f30b2c7d17c064e7e18856208b364bf53f0e9e1684a1165dd6f5ca94d98a5d01473044022059e9639cd64fe0c25c49a81993871aea9b5be81cf50d3507a685d84b5d2e153702205f4fb28081096f24f00686d23819d75f3c7f288dec7f9622d38b8d2172a63f710169522102952139ccb86b16383b29039bf76febbbfb1c41c15ec631c1b13f1424f7680a9321021c44aa55510b353e397f9f2b3e7dee36282f1421651fef96c31ba612c11627462103bbf6336d9f021ae4ee642af75156b644195dbe23a8de5cb62600b58e139d242653ae0400473044022070c844b19f917c16fc644957bafb3308b2cb244cdbce48ab763302d290272ef002204f6ca015707a9ffcbc3bee63c66a035430a7988495ce2f9c1d8e22f81c9ab0890147304402200ece13be1073c262d5a86c8078ce16fe6f2fa456665e6284eddd42b42006f74e022074fa8226a4ca112483736a9f0b47800b99197ffc8df3bcce6c3bef268a2a526a0169522102216a2779be169379d783d25180040e079da9193770e63bd5de8dc06c5834c0bd2102da354fb8873567ae83fef73a4fa783448a9b481e7f69619a20983149b1f19b402102e64c3c1e222b9cbdfb7f70b5bb011b89e0642eb7cfc94f2750948aeb54f8046553ae04004730440220431cde3ecb90c38c7f1bbafb4a417a8183e314eecc8a374fc67d5847b6822fd102201ff9e2ecc57bcdfb69ba1a9f78d132d7123b37e6838c7e77e37278b6269384e301483045022100df2ba7d95b973d30871a90f35757ad13b61c9f2deb8e36b675d19de4d5dfd12f02203f3b5dadd71304080a63c4031f4a6d7b30f3f742fc44e392a69cca947159d0890169522103630ae85a979ee329fa729b361a114254da4e871bce8629e39c26a26fb6ab85e82102d566976b3c356c2062a75e938ee538edef5ced6f6f8ca827c3b004b6c5f256a42102f5de580abf9fd69275cfc105ba165664d1d79b06f3aae1f512d778624c9ef8cb53ae0400473044022064b121d06f145acd2e84cf6ae73c6c8ba8119ac9450a93c9a8fcb882c98a7bd102206a5e8ff6b4b16b212d3c1cc92864ef9b2c804bc26f268399799a36414664923c0147304402200a99894da1fad45c7d62aceb0aa9d7a05f4304e4498327d8827f308ed8e70f43022034f68f701c3bb2afedc69ab7ba402a0d2277ef530fd89376836bb55151cf50540169522103ab69777a57f9cce9f3eea8079a2a1c89c07f9068022bf80686242efa1b60bbe021023718e1ca5f2f6e38841b627a8fa5c9a66f767d6a05367508cea960f21bb2feea2102e56e70e1cfac75b661131ac924aee909920d948ffb08f879bfa5ea1d3fecb9d353ae04004830450221008244512d55c6841c77a4d126fa5a33c92c247bf61d47c7dca0a7256ae8212e5f022019838e0aa334efa4a590b7398f3fdf6c2ef2a3d00f3856814ed04a9a0e56e54c01483045022100a8972a7a9a6852b0f96de6fa0fba67748551591ff9bec42a0df7d1f0a03392670220254d76e53a9c777aedbd2937c2f45324ff88a3ea15103bf9e325bd336d6cb2f501695221027d0f541fdf168131c9003d824c486062a9b6869f000cd65639a76255d05c3ee72103e86ad41675d5fcb7ce4d4b60e8f2ad122f873239d514bf1bceca49b74536a8aa2102805e8f635752cf9a42c38bb80f224b3212882e6c3c50bebba52eb7ab0935cae653ae040047304402202c71af53522c21e981863a336fd8b98394a981dbf8ac4a75c7b4f48c924faf56022057ef76dfcb327d39b4bc79ca5d3d604522ecfae6ee434f943ea714cf5ba0593301473044022068282d81f9968ef31db541e9b72aefff62aefd889e12c97f61e41e20d57d715d02202f54462beb8c4a462583c4917c3d937c036c5c5983c84c5be132cb9bef8d908c0169522102f116a19e2900371ae7f4be5523d9845654dedbce090e2fa30e816732d5d8251d2102d6f4005dfc28cb3c0def8c1f52f2d23d486dfbfa3ffa43c94a4b3d43c920c93e2103c6fac560735d78d95cd63652675e643c8e0433002679cd032143d8411dba6b1153ae0400483045022100a6bdc5206380571da68cde53f1ef19dca58c4d2d9cbc4220fe7e12e70fc9b0bd022011cad80c2bf83f85c111d5d3891f634a393c7ca57513ff2b5813e87bc85ae50501473044022100f598f696af768e2290f46873c3873876659fe60fe73c0068136e746a8f3327ef021f1b9de6928c8cf1245605ae705d7d1304efbbd385d4b77a81ab66bf8dda6a9401695221038acc7e5bd34e1198341f6c5561eba4580ce930ebfd78e9705fa51aaa34b8035b2103a40b87c3a637f4ab9b4cfadcf0537bfb71753b4f070e896eeeb0615bb21024f12103dbbc3ab658aef4b292042c3e99de6affcf573249a40bd6c225e4d84168bd01b253ae0400473044022041661469f59e7a72a9e1835323397fd4920e3a234c5c346afe4974c589ae25ee02207fe4b1c42b13dec0e1372446db604f8bda2be95c9bf0aa2c26dad8530ad11170014730440220417ce3b947dc37368af439c0afd347bcbbfed55b769e29bed1947079671197bf022079747806204ce6828a8d6b34703781f049124c39f8385fb1d1a880d9325e98850169522102aee5147d14dc3aa1da07594703864035112c1ce140ea7955599e9c3b911d65bc21022942c8ffc8271de9e763a3299bfc919b4077455c010aec920abd085df6ab73a621022c3f85c185fd927ed3e81f02ccaae1d45720c0929a0d1a6e9accd0c8d04342ef53ae040047304402206710fd815cbc1aaefae6fbc4b603f4f8b96516ab9f5dd69f371692ff0c2bfd34022040c6aebb39386aff976bb468b0c71e350a378b5a3847ccc9947ee522bb687c4e0147304402202cf317b968ed838f19634f5ea56ec0a521bbefeffe7a4c9498899d2b23fff4d5022009610c7bbd95f7cb212ee07299edfbb56c6ac8bf796b14ae9f2f64d5a08b79e30169522103ab69777a57f9cce9f3eea8079a2a1c89c07f9068022bf80686242efa1b60bbe021023718e1ca5f2f6e38841b627a8fa5c9a66f767d6a05367508cea960f21bb2feea2102e56e70e1cfac75b661131ac924aee909920d948ffb08f879bfa5ea1d3fecb9d353ae00000000", "id": 17}
>>>>>>>>
Stack trace:
  at Microsoft.FSharp.Core.PrintfModule+PrintFormatToStringThenFail@1379[TResult].Invoke (System.String message) [0x00001] in <59964427904cf4daa745038327449659>:0 
  at TcpEcho.StratumClient.Deserialize[T] (System.String json) [0x00030] in /Users/knocte/Documents/Code/tcpEchoLuke/src/StratumClient/StratumClient.fs:39 
  at <StartupCode$StratumClient>.$StratumClient+BlockchainTransactionGet@64-1.Invoke (System.String _arg2) [0x00002] in /Users/knocte/Documents/Code/tcpEchoLuke/src/StratumClient/StratumClient.fs:64 
  at Microsoft.FSharp.Control.AsyncBuilderImpl+args@823-1[a,b].Invoke (a a) [0x00016] in <59964427904cf4daa745038327449659>:0 
--- End of stack trace from previous location where exception was thrown ---

  at Client.Tests.ClientTest.ConnectWithElectrumServersTransactionGet () [0x00060] in /Users/knocte/Documents/Code/tcpEchoLuke/src/Client.Tests/ClientTests.cs:42 
  at NUnit.Framework.Internal.ExceptionHelper.Rethrow (System.Exception exception) [0x00006] in <64cd73ddfc8a470aa513c8243b429bdb>:0 
  at NUnit.Framework.Internal.AsyncInvocationRegion+AsyncTaskInvocationRegion.WaitForPendingOperationsToComplete (System.Object invocationResult) [0x00030] in <64cd73ddfc8a470aa513c8243b429bdb>:0 
  at NUnit.Framework.Internal.Commands.TestMethodCommand.RunAsyncTestMethod (NUnit.Framework.Internal.TestExecutionContext context) [0x00038] in <64cd73ddfc8a470aa513c8243b429bdb>:0 
```

I wonder if this happens when the response is bigger than the hardcoded buffer length of 1024?

lukethenuke committed:
Fix garbled response and timeout error handling

lukethenuke commented:
Hard bug to track down, but hopefully easy fix :)

lukethenuke commented:
> When running the test I caught a NRE, which I think is because in some cases the IncompleteResponse is really not wrapping anything of value. 

This happens because JSON.NET serialize empty input strings to ```null```,  i have added a check to the IncompleteResponseException error handling which converts it into ConnectionUnsuccessfulException as you requested. 
If the server responds with an empty response (and closes gracefully, no timeout) then we will have the same problem. Should we handle this or?

knocte commented:
> Should we handle this or?

Oh yes please!

BTW I've just realised that maybe the wrapping with AggregateException happens because you're using Task API now? I mean Task.WhenAny; I think it can be converted to this which is a bit cleaner (and maybe makes the AggregateException wrapping disappear): https://stackoverflow.com/a/20521059/544947

lukethenuke committed:
Fix error handling for empty json responses

lukethenuke commented:
The reason for using Tasks as the timeout is because I could not get it to correctly timeout with Async, either with StartChild or the code you linked. I do not know why.

knocte commented on src/StratumClient/StratumClient.fs:
```diff
@@ -29,6 +30,7 @@ type StratumClient(endpoint: string, port: int) =
     inherit Client(endpoint, port)
 
     static member Deserialize<'T> json: 'T =
+        if String.IsNullOrWhiteSpace(json) then failwith "Invalid json format, the server response was empty"
```
shouldn't you throw an ex that inherits from ConnectionUnsuccessfulException? otherwise this would break the tests

lukethenuke replied:
I saw this as an edge case so it should break hard, but sure that makes sense. Maybe anything thrown from the deserialization should be wrapped in a ConnectionUnsuccessfulException (or derivative)?

knocte replied:
actually you're right, if we break hard for malformed json, we should break hard here too

lukethenuke replied:
So you want to wrap all deserialization errors or let it be?

knocte replied:
let it be

knocte commented:
> The reason for using Tasks as the timeout is because I could not get it to correctly timeout with Async, either with StartChild or the code you linked. I do not know why.

Maybe because of using Task.Delay instead of Async.Sleep?

lukethenuke commented:
> Maybe because of using Task.Delay instead of Async.Sleep?

No, using Async.Sleep, example:
```fsharp
let withTimeout (timeout : int) (xTask: Task<'a>) = async {
    let delay = async {
        do! Async.Sleep(timeout)
        return None
    }
    let value = async {
        let! res = xTask |> Async.AwaitTask
        return Some res
    }
    let! choice = Async.Choice [value; delay]
    match choice with
    | None -> return Timeout
    | Some x -> return Result x
}
```

lukethenuke committed:
Replace Async.Catch with try-with

lukethenuke committed:
Replace mkString implementation to use StringBuilder instead

lukethenuke committed:
Change withTimeout implementation to use Async.Choice

lukethenuke committed:
Change socket send to the Async version

lukethenuke committed:
Add timeout to socket send

knocte commented on src/Client/Client.fs:
```diff
@@ -0,0 +1,119 @@
+                return! readFromPipeAsync reader totalString
+        with
+            // If we got an TimeoutException anywhere in the exception chain
+            | :? AggregateException as ae when ae.Flatten().InnerExceptions |> Seq.exists (fun x -> x :? TimeoutException) ->
```
now that we're not using Task.WhenAny or Task.Delay, can you double check if exceptions still come wrapped under AggregateException?

lukethenuke replied:
Since it is ```reader.ReadAsync()``` that throws the exception it will be wrapped in an AggregateException.
1) The last rethrow case is not be needed anymore however (aswell as the rethrow in StratumClient#Call).
2) Should the readFromPipeAsync also use the StringBuilder instead of concating strings?

knocte replied:
1. ok then let's remove them
2. yes please

but, hey, and sorry for the delay, but I rebased your PR somewhere else, because I wanted a CI pipeline to run it, and unfortunately it hangs, so we should fix that first: https://gitlab.com/nblockchain/TcpEchoSharp/merge_requests/1 , let me know your gitlab username and I'll give you access to the repo that holds the PR
